### ISSUE TITLE
Racing working-set updates cause "Clear working set query" errors

### ISSUE THREAD

**Issue by djmitche** [GothenburgBitFactory/taskwarrior#3325]
Details in #3323.

A good place to start here would be to reproduce this issue: create a task db with a bunch of tasks in it and then run a loop displaying `task` output in such a way that it will clear and rebuild the working set every time. It will probably be necessary to run several such loops at the same time (e.g., in different terminal windows).

Once it's reproduced, the next step is to figure out what the actual error is, and maybe why that's not being shown.

Ideally Taskwarrior's DB should be resilient to concurrent access, but perhaps this is a transaction-failure error that occurs when one process writes entries into the working set concurrently with another process executing `delete from working_set`. If so, perhaps we can catch and re-try that transaction.

**Comment by djmitche** [comment]
Indeed, running two reports concurrently in a loop eventually results in
```
Clear working set query: database is locked: Error code 5: The database file is locked
```

One issue is, we're using the old (2010!) rollback journal mode and not WAL. However, just using WAL is not enough - I still see the error, and indeed https://www.sqlite.org/wal.html says
> 9. Sometimes Queries Return SQLITE_BUSY In WAL Mode
> 
> The [second advantage of WAL-mode](https://www.sqlite.org/wal.html#advantages) is that writers do not block readers and readers to do not block writers. This is mostly true. But there are some obscure cases where a query against a WAL-mode database can return [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy), so applications should be prepared for that happenstance.
>
> Cases where a query against a WAL-mode database can return [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) include the following:
>
> * If another database connection has the database mode open in [exclusive locking mode](https://www.sqlite.org/pragma.html#pragma_locking_mode) then all queries against the database will return [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy). Both Chrome and Firefox open their database files in exclusive locking mode, so attempts to read Chrome or Firefox databases while the applications are running will run into this problem, for example.
> * When the last connection to a particular database is closing, that connection will acquire an exclusive lock for a short time while it cleans up the WAL and shared-memory files. If a separate attempt is made to open and query the database while the first connection is still in the middle of its cleanup process, the second connection might get an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) error.
> * If the last connection to a database crashed, then the first new connection to open the database will start a recovery process. An exclusive lock is held during recovery. So if a third database connection tries to jump in and query while the second connection is running recovery, the third connection will get an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) error.

So some additional work is required..

There's some support for automatically retrying on a busy signal (https://docs.rs/rusqlite/latest/rusqlite/struct.Connection.html#method.busy_timeout) but that seems not to be in effect here.

**Comment by djmitche** [comment]
Ah, that's because we were using BEGIN DEFERRED (the default). See https://www.sqlite.org/lang_transaction.html. DEFERRED means that when the first statement is executed, if it's SELECT then a read-only transaction is begun. This is then upgraded to a write transaction if a later query performs a write. However, it seems that the busy handler doesn't handle this situation (perhaps because it could lead to deadlock?).

Switching to BEGIN IMMEDIATE (and using WAL mode) fixes the problem. That might complicate a later read-only mode, but that's a problem for another issue.

https://github.com/GothenburgBitFactory/taskchampion/pull/412 should address this automatically once Taskwarrior is updated to include it.

**Comment by djmitche** [comment]
As noted in #3676, this did not get fixed in 3.1.0, which uses TaskChampion 0.6.0. It will be in 3.2.0.
